### PR TITLE
ESQL: Speed up CASE for some parameters

### PR DESCRIPTION
**Pull request by nik9000**
This speeds up the `CASE` function when it has two or three arguments and both of the arguments are constants or fields. This works because `CASE` is lazy so it can avoid warnings in cases like
```
CASE(foo != 0, 2 / foo, 1)
```

And, in the case where the function is *very* slow, it can avoid the computations.

But if the lhs  and rhs of the `CASE` are constant then there isn't any work to avoid.

The performance improvment is pretty substantial:
```
 (operation)  Before   Error   After    Error  Units
 case_1_lazy  97.422 ± 1.048  101.571 ± 0.737  ns/op
case_1_eager  79.312 ± 1.190    4.601 ± 0.049  ns/op
```

The top line is a `CASE` that has to be lazy - it shouldn't change. The 4 nanos change here is noise. The eager version improves by about 94%.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java:75]: I *suppose* this is something that we could derive from the Expression - but it felt simpler to put it here. And it's just a boolean at this point - though maybe it should be a cost estimate at some point.

**Comment by nik9000** [comment]
I'm waiting until I merge #112401 to pick this one back up.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java:301]: I can't tell if the logic is wrong here or if the naming is just confusing.  To me, this reads like "if it is safe to evaluate in lazy, return an eager evaluator".  I would normally say "eager evaluation" was the opposite of "lazy evaluation".

Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java:301]: Bad name, I think. It's "if this is safe to eagerly evaluate when we expect a lazy"

Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java]: I'll rename this now that the CASE bugs are fixed.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java
Comment (costin) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java:76]: In case of constants, the unvisited branches can be removed and the case simplified.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java
Comment (costin) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java:74]: another alternative is to be extract this as a marking interface that gets implemented by certain factories.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/EvalOperator.java:76]: That works if the expression is constant, but not if the *values* are constant. We still have to evaluate in that case. And when we do we can do it the fast way with something like this.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java
Comment (luigidellaquila) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java:513]: We could leverage `BooleanVector.isConstant()` here, eg.

```
                BooleanVector mask = lhsOrRhs.mask();
                if (mask.isConstant()) {
                    builder.copyFrom(mask.getBoolean(0) ? lhs : rhs, 0, lhs.getPositionCount());
                } else {
                    for (int p = 0; p < lhs.getPositionCount(); p++) {
                    ...
```
Or even return lhs/rhs...? Not sure it's safe for memory accounting

**Review by luigidellaquila — APPROVED** [review]
LGTM
The method names are a bit cryptic indeed, but the javadoc helps a lot

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java:513]: I believe it is safe, but tests can figure it out. I'll try it.

**Pull request merged**
